pkg/operator/resource/resourceapply/monitoring.go

```diff
@@ -2,148 +2,150 @@ package resourceapply
+	return ApplyUnstructuredResourceImproved(ctx, client, recorder, required, noCache, prometheusRuleGVR)
+}
 
-// equalityChecker allows for custom equality comparisons.  This can be used to allow equality checks to skip certain
-// operator managed fields.  This capability allows something like .spec.scale to be specified or changed by a component
-// like HPA.  Use this capability sparingly.  Most places ought to just use `equality.Semantic`
-type equalityChecker interface {
-	DeepEqual(a1, a2 interface{}) bool
+// DeletePrometheusRule deletes the PrometheusRule.
+func DeletePrometheusRule(ctx context.Context, client dynamic.Interface, recorder events.Recorder, required *unstructured.Unstructured) (*unstructured.Unstructured, bool, error) {
+	return DeleteUnstructuredResource(ctx, client, recorder, required, prometheusRuleGVR)
 }
 
-// ApplyServiceMonitor applies the Prometheus service monitor.
+// ApplyServiceMonitor applies the ServiceMonitor.
 func ApplyServiceMonitor(ctx context.Context, client dynamic.Interface, recorder events.Recorder, required *unstructured.Unstructured) (*unstructured.Unstructured, bool, error) {
+	return ApplyUnstructuredResourceImproved(ctx, client, recorder, required, noCache, serviceMonitorGVR)
+}
+
+// DeleteServiceMonitor deletes the ServiceMonitor.
+func DeleteServiceMonitor(ctx context.Context, client dynamic.Interface, recorder events.Recorder, required *unstructured.Unstructured) (*unstructured.Unstructured, bool, error) {
+	return DeleteUnstructuredResource(ctx, client, recorder, required, serviceMonitorGVR)
+}
+
+// ApplyUnstructuredResourceImproved can utilize the cache to reconcile the existing resource to the desired state.
+func ApplyUnstructuredResourceImproved(ctx context.Context, client dynamic.Interface, recorder events.Recorder, required *unstructured.Unstructured, cache ResourceCache, resourceGVR schema.GroupVersionResource) (*unstructured.Unstructured, bool, error) {
+	name := required.GetName()
 	namespace := required.GetNamespace()
-	existing, err := client.Resource(serviceMonitorGVR).Namespace(namespace).Get(ctx, required.GetName(), metav1.GetOptions{})
+
+	// Create if resource does not exist, and update cache with new metadata.
+	existing, err := client.Resource(resourceGVR).Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
 	if errors.IsNotFound(err) {
-		newObj, createErr := client.Resource(serviceMonitorGVR).Namespace(namespace).Create(ctx, required, metav1.CreateOptions{})
-		if createErr != nil {
-			recorder.Warningf("ServiceMonitorCreateFailed", "Failed to create ServiceMonitor.monitoring.coreos.com/v1: %v", createErr)
-			return nil, true, createErr
-		}
-		recorder.Eventf("ServiceMonitorCreated", "Created ServiceMonitor.monitoring.coreos.com/v1 because it was missing")
-		return newObj, true, nil
+		requiredCopy := required.DeepCopy()
```

Comment:
why do we need a deepCopy here? I can't see any call that may alter it.

Reply:
Dropped.